FEM-2357 Add External (out-of-band) WebVTT captions support

x-NR-x opened the pull request:
Supports external subtitles for media which has no internal subtitles.

- Moved AssetHandler and DefaultAssetHandler under the same folder named AVAsset.
- Added a PKAssetResourceLoaderDelegate to hold all optional delegates.
- Updated the DefaultAssetHandler to have a PKAssetResourceLoaderDelegate.
The avAsset will always have the PKAssetResourceLoaderDelegate when set.
Code was updated accordingly for the FP.
- Added a PKCaptionsAssetResourceLoaderDelegate to support external captions.
Added a PKExternalSubtitle class to provide to the PKCaptionsAssetResourceLoaderDelegate.

x-NR-x committed:
FEM-2357 Moved AssetHandler and DefaultAssetHandler under the same folder named AVAsset.

x-NR-x committed:
FEM-2357 Added a PKAssetResourceLoaderDelegate to hold all optional delegates.

x-NR-x committed:
FEM-2357 Updated the DefaultAssetHandler to have a PKAssetResourceLoaderDelegate.
The avAsset will always have the PKAssetResourceLoaderDelegate when set.
Code was updated accordingly for the FP.

x-NR-x committed:
FEM-2357 Added a PKCaptionsAssetResourceLoaderDelegate to support external captions.
Added a PKExternalSubtitle class to provide to the PKCaptionsAssetResourceLoaderDelegate.

x-NR-x committed:
FEM-2357 Supports external subtitles for media which has none.

x-NR-x committed:
FEM-2357 Removed prints to the console, added PKLog.

x-NR-x committed:
FEM-2357 Added the option to set the characteristics tag.

x-NR-x committed:
FEM-2357 Added AutoSelect to the available tags.
Updated the initializer to support a null value in Objc for the String defaults in Swift.

noamtamim commented:
AssetResourceLoaderDelegate refactor:
https://github.com/kaltura/playkit-ios/compare/develop...93eb3621de928f554e94e5bcfe41be5d45133a78

Captions feature:
https://github.com/kaltura/playkit-ios/compare/93eb3621de928f554e94e5bcfe41be5d45133a78...FEM-2357

noamtamim commented on Classes/Player/AVAsset/DefaultAssetHandler.swift:
```diff
@@ -14,16 +14,32 @@ import PlayKitUtils
     }
     
+    private func replaceURL(_ url: URL, withScheme scheme: String) -> URL {
+        let urlString = url.absoluteString
```
Consider using URLComponents for the replacement.

x-NR-x replied:
Yes, will do. 
I copied it from the sample.

noamtamim commented on Classes/Player/AVAsset/DefaultAssetHandler.swift:
```diff
@@ -50,10 +66,22 @@ class DefaultAssetHandler: AssetHandler {
             return
         }
 
+        // Set the custom scheme for the external subtitles, if exists.
```
Does it have to be in this class?

x-NR-x replied:
Yes, cause I need to do it before creating the AVURLAsset.

noamtamim commented on Classes/Player/Tracks/PKCaptionsAssetResourceLoaderDelegate.swift:
```diff
@@ -0,0 +1,207 @@
+        return allSubtitles
+    }
+    
+    func getSubtitlem3u8(forSubtitle subtitle: PKExternalSubtitle) -> String {
```
Consider moving this function to PKExternalSubtitle (as `buildM3u8Playlist()` or similar).

x-NR-x replied:
Yes, you'r right, better approach. Will change it.

noamtamim commented on Classes/Player/Tracks/PKCaptionsAssetResourceLoaderDelegate.swift:
```diff
@@ -0,0 +1,207 @@
+        return true
+    }
+    
+    func getSubtitlesEXT() -> String {
```
Consider moving this function to PKExternalSubtitle (as `buildMasterLine()` or similar).

x-NR-x replied:
I will create an additional function in the PKExternalSubtitle as buildMasterLine().
But this function will still exist cause it goes over all the external subtitles.

x-NR-x committed:
FEM-2357 Creating the m3u8 playlist and master line inside the PKExternalSubtitle class.
Updated the function for replacing the url scheme via URLComponents.

noamtamim commented on Classes/Player/Tracks/PKCaptionsAssetResourceLoaderDelegate.swift:
```diff
@@ -16,7 +16,6 @@ class PKCaptionsAssetResourceLoaderDelegate: NSObject, AVAssetResourceLoaderDele
     
     private let extM3UPrefix = "#EXTM3U"
```
I don't think these constants should be instance fields. Either static fields or in the file itself, as `fileprivate` (preferred).

x-NR-x replied:
Created a struct with all the ones we are using.

noamtamim commented on Classes/Player/Tracks/PKExternalSubtitle.swift:
```diff
@@ -0,0 +1,162 @@
+     
+        - characteristics:  The value contains one or more Uniform Type Identifiers [UTI] separated by comma (,) characters. This attribute is OPTIONAL. Each UTI indicates an individual characteristic of the Rendition.
+     
+        - vttURLString:     This is the string URL of the vtt.
```
Why is it optional? How does it work if not specified?

x-NR-x replied:
vttURLString is not optional.
characteristics is optional.

noamtamim replied:
It's the most important parameter, it should be moved to before the optional ones.
`init(id: String, name: String, language: String? = "", isDefault: Bool = false, autoSelect: Bool = false, forced: Bool = false, characteristics: String? = "", vttURLString: String, duration: Double)`

It's confusing that vttURLString and duration that don't have default values appear AFTER parameter that do have defaults.

noamtamim commented on Classes/Player/Tracks/PKExternalSubtitle.swift:
```diff
@@ -0,0 +1,162 @@
+     
+        - forced:           A value of true indicates that the Rendition contains content that is considered essential to play. When selecting a FORCED Rendition, a client SHOULD choose the one that best matches the current playback environment (e.g., language). This attribute is OPTIONAL. It's absence indicates an implicit value of false.
+     
+        - language:         The value is a one of the standard Tags for Identifying Languages [RFC5646], which identifies the primary language used in the Rendition. This attribute is OPTIONAL.
```
Should be a mandatory (but `Optional`) parameter. And it should be before the other parameters.

noamtamim commented on Classes/Player/Tracks/PKExternalSubtitle.swift:
```diff
@@ -0,0 +1,162 @@
+    }
+    
+    public override var description: String {
+        return super.description + """
```
Maybe just return the master line as description?

x-NR-x replied:
I can, but I will also want to see the value of the given vttURLString and the given duration.

x-NR-x committed:
FEM-2357 Changed relative paths to absolute paths.
Added a check, if internal subtitles exist, not to add the external subtitles.
Set functions to private where needed.
Added a line break also before the addition of the external subtitles.

x-NR-x committed:
FEM-2357 Set the Playlist tags as a struct to hold all tags we use.
Code cleanup.

x-NR-x committed:
FEM-2357 Set the, PKExternalSubtitle, language property right after the name property.

x-NR-x committed:
FEM-2357 Set the language to non optional.
Reordered the vttURLString and the duration.
